File: 188. Best Time to Buy and Sell Stock IV/188. Best Time to Buy and Sell Stock IV.java
class Solution {
    public int maxProfit(int k, int[] prices) {
        int n=prices.length;
        int maxDeal=k*2;
        int dp[][]=new int[maxDeal][n];
        
        for(int d[]:dp) Arrays.fill(d,Integer.MIN_VALUE);
        int m = Integer.MIN_VALUE;
        for (int i = 0; i < n; i++) {
            m = Math.max(m, -prices[i]);
            dp[0][i] = m;
        }
        for (int i = 1; i < maxDeal; i++) {
            for (int j = i; j < n; j++) {
                if(i%2 == 0) {
                    dp[i][j] = Math.max(dp[i][j-1], dp[i-1][j-1] - prices[j]);
                }
                else{
                    dp[i][j] = Math.max(dp[i][j-1], dp[i-1][j-1] + prices[j]);
                }
            }
        }
        
        int ans = 0;
        for(int i = maxDeal-1; i >= 0; i--){
            ans = Math.max(ans, dp[i][n-1]);
        }
        
        return ans;
    }
}

File: 188. Best Time to Buy and Sell Stock IV/188. Best Time to Buy and Sell Stock IV.cpp
class Solution {
public:
    int maxProfit(int k, vector<int>& prices) {
        int days = prices.size(), tran = min(k*2, days);
        vector<vector<int>> dp(tran, vector<int>(days,0));
        dp[0][0] = -prices[0];
        for(int i = 1; i < days; ++i){
            dp[0][i] = max(dp[0][i-1], -prices[i]);
        }
        for(int i = 1; i < tran; ++i){
            int p = (i & 1) ? 1 : -1;
            dp[i][i] = dp[i-1][i-1] + p*prices[i];
            for(int j = i+1; j < days; ++j){
                dp[i][j] = max(dp[i][j-1], dp[i-1][j-1]+p*prices[j]);
            }
        }
        int ans = 0;
        for(auto& vec: dp){
            ans = max(ans, vec.back());
        }
        return ans;
    }
};